bftengine/src/preprocessor/messages/PreProcessResultMsg.cpp

```diff
@@ -70,7 +70,15 @@ std::optional<std::string> PreProcessResultMsg::validatePreProcessResultSignatur
   }
 
   auto hash = concord::util::SHA3_256().digest(requestBuf(), requestLength());
+  std::unordered_set<bftEngine::impl::NodeIdType> seen_signatures;
   for (const auto& s : sigs) {
+    const auto [it, created] = seen_signatures.insert(s.sender_replica);
+    (void)it;  // unused variable
+
+    if (!created) {
```

Comment:
I have question:
Consider that 2 signatures are same. Then this is an error case and it will return after that.
Now will the verifySig() can return true for this signed hash for another replica. 
I think same signature with different ID will not be an issue.
But if the tuple s.signature and s.sender_replica [i.e both the signature and replica id] are same then this will be an issue.
So say a byzantine primary sends 7 same signature for his own hash with same sender_replica ids (ie. his own replica id). Then only this will be an issue.
In rare (rarest) case due to collision, signature of two different replica can also be same. So in such case we will get a false issue due to non checking of replicaid alongwith the signatures.

Reply:
Sorry @arc-vmware but I don't follow.
We keep the sender id in the set. The purpose of this check is to ensure that all sender IDs are unique. After that we verify each signature with the public key of the corresponding replica.
Even if the signatures are the same (which should not happen) everything will be fine as long as the signature verification works.
As far as I understand - this logic should be fine. What I am missing?

Reply:
Yes @tdimitrov , my bad, I assumed that signatures are checked. But as you are checking replica id, its fine. Thanks for correcting me.